### PR TITLE
Add isolayer selection method

### LINKED ISSUE #3845: Isolayer selection
## Is your feature request related to a problem? ##
There are currently methods of selecting a spherical or cylindrical "shell" where atoms are at least r_min away but less than r_max away, however it would be nice if there was something for flexible.

## Describe the solution you'd like ##
Such an isolayer selection tool could:
1) find a flexible shell region relative to a polymer backbone, where it is sort of cylindrical but not quite
2) Select the second hydration shell of ions, while excluding water than is in the first shell of another ion.

### PR DESCRIPTION
**Pull request by jaclark5**
Fixes #3845

Changes made in this Pull Request:
 - Added isolayer selection method

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [X] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

**Comment by jaclark5** [comment]
This selection method was used in a publication soon to be submitted.

File: testsuite/MDAnalysisTests/core/test_atomselections.py [hunk 1]
```diff
@@ -689,6 +697,34 @@ def test_spherical_layer(self, u, periodic):
...
+        cog1 = r1[0].position.reshape(1, 3)
+        d1 = distance_array(u.atoms.positions, cog1, box=box)
+        cog2 = r1[1].position.reshape(1, 3)
+        d2 = distance_array(u.atoms.positions, cog2, box=box)
+
+        ref_inner  = set(np.where((d1 < rmin) | (d2 < rmin))[0])
+        ref_outer  = set(np.where((d1 < rmax) | (d2 < rmax))[0])
+        ref_outer -= ref_inner
+
+        assert ref_outer == set(result.indices)
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/core/test_atomselections.py]: just to check, this selection is capturing something right?

Comment (jaclark5) [reply testsuite/MDAnalysisTests/core/test_atomselections.py]: Hey! Yes, there are about 12 or so atoms. Do you want me to add something with the length?

**Review by richardjgowers — APPROVED** [review]
This looks good to me

**Comment by richardjgowers** [comment]
@jaclark5 since this PR is likely to be the first merged, can you add yourself to `package/AUTHORS` too?

**Review by richardjgowers — DISMISSED** [review]
Oops, actually we've got docs for selections also here: https://github.com/MDAnalysis/mdanalysis/blob/develop/package/doc/sphinx/source/documentation_pages/selections.rst

Can you include your new selection on that doc page too?

File: package/doc/sphinx/source/documentation_pages/selections.rst [hunk 2]
```diff
@@ -245,6 +245,14 @@ sphzone *externalRadius* *selection*
     resid 130 or resid 80 ) )`` selects the center of geometry of protein,
     resid 130, resid 80 and creates a sphere of radius 6.0 around the COG.
 
+isolayer *inner radius* *outer radius* *selection*
+    Similar to sphlayer, but will find layer around all referenced atoms. 
+    For example, if the atom types for a polymer backbone were chosen, then
+    an isolayer parallel to the backbone will be generated. As another 
+    example, if an ions were chosen as a reference to isolate the second
```
Comment (hmacdope) [inline package/doc/sphinx/source/documentation_pages/selections.rst]: Did you mean a single or multiple ions?
```suggestion
    example, if an ion was chosen as a reference to isolate the second 

```

File: package/doc/sphinx/source/documentation_pages/selections.rst [hunk 3]
```diff
@@ -245,6 +245,14 @@ sphzone *externalRadius* *selection*
     resid 130 or resid 80 ) )`` selects the center of geometry of protein,
     resid 130, resid 80 and creates a sphere of radius 6.0 around the COG.
 
+isolayer *inner radius* *outer radius* *selection*
+    Similar to sphlayer, but will find layer around all referenced atoms. 
+    For example, if the atom types for a polymer backbone were chosen, then
+    an isolayer parallel to the backbone will be generated. As another 
+    example, if an ions were chosen as a reference to isolate the second 
+    hydration layer, then those solvent atoms that overlap will not include
```
Comment (hmacdope) [inline package/doc/sphinx/source/documentation_pages/selections.rst]: I'm not sure I fully follow the use of the word "overlapping" here? Could you clarify for me possibly?

File: package/MDAnalysis/core/selection.py [hunk 4]
```diff
@@ -356,6 +356,48 @@ def _apply(self, group):
...
+            return sys[[]]
+
+        box = group.dimensions if self.periodic else None
+        pairs_outer = distances.capped_distance(sel.positions, sys.positions,
+                                                self.exRadius, box=box,
+                                                return_distances=False)
+        pairs_inner = distances.capped_distance(sel.positions, sys.positions,
+                                                self.inRadius, box=box,
+                                                return_distances=False)
+        inner = np.array(pairs_inner).T[1]
```
Comment (hmacdope) [inline package/MDAnalysis/core/selection.py]: As far as I can tell these two variables are not used again? perhaps I am missing something.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looks really good! Just a few quick comments.

File: package/doc/sphinx/source/documentation_pages/selections.rst (hunk 2 above)
Comment (jaclark5) [reply package/doc/sphinx/source/documentation_pages/selections.rst]: I meant multiple. I've pushed the changes.

File: package/doc/sphinx/source/documentation_pages/selections.rst (hunk 3 above)
Comment (jaclark5) [reply package/doc/sphinx/source/documentation_pages/selections.rst]: I think I clarified it adequately in my latest push.

File: package/MDAnalysis/core/selection.py (hunk 4 above)
Comment (jaclark5) [reply package/MDAnalysis/core/selection.py]: You're right! Sorry about that.

File: package/CHANGELOG [hunk 5]
```diff
@@ -25,6 +26,7 @@ Fixes
   * fixes guessed_attributes and read_attributes methods of the Topology class, as 
     those methods were giving unexpected results for some topology attributes
     (e.g. bonds, angles) (PR #3779).
+  * Added isolayer selection method (Issue #3845)
```
Comment (hmacdope) [inline package/CHANGELOG]: Could you please change this to newest first? Other than that looks good to go pending CI.

**Review by hmacdope — CHANGES_REQUESTED** [review]
One tiny thing then good to go.

File: package/CHANGELOG (hunk 5 above)
Comment (jaclark5) [reply package/CHANGELOG]: Oh, I didn't realize, thanks!

File: package/CHANGELOG [hunk 6]
```diff
@@ -13,19 +13,20 @@ The rules for this file:
...
 Fixes
+  * Added isolayer selection method (Issue #3845)
   * Auxiliary; determination of representative frames: Removed undefined
     behaviour for cutoff values < -1 (PR # 3749)
   * Fixes distances.between not always returning AtomGroup (Issue #3794)
   * Upgrade to chemfiles 0.10.3 (Issue #3798)
   * fixes guessed_attributes and read_attributes methods of the Topology class, as 
     those methods were giving unexpected results for some topology attributes
     (e.g. bonds, angles) (PR #3779).
-
```
Comment (hmacdope) [inline package/CHANGELOG:28]: Again so sorry for the nitpick, but can you add this blank line back in :)

**Review by hmacdope — CHANGES_REQUESTED** [review]
Sorry one more nitpick I promise.

**Review by hmacdope — APPROVED** [review]
Congrats @jaclark5! Thanks for the contribution!

**Comment by hmacdope** [comment]
@jaclark5 now that you have merged your first PR (yay) would you be able to please introduce yourself on the developer mailing list (https://groups.google.com/g/mdnalysis-devel)? This is so we have a traceable email for all authors and know a bit more about you. It is also important as we are in the slow process of re-licensing.

**Pull request merged**